datagateway_api/src/search_api/query_filter_factory.py

```diff
@@ -0,0 +1,151 @@
+                    raise FilterError(
+                        "No valid filter name given within filter query param",
+                    )
+        elif query_param_name == "where":
+            # For the count endpoints
+            query_filters.extend(
+                SearchAPIQueryFilterFactory.get_query_filter(
+                    {"filter": request_filter},
+                ),
+            )
+        else:
+            raise FilterError(
+                f"Bad filter, please check query parameters: {request_filter}",
+            )
+
+        return query_filters
+
+    @staticmethod
+    def get_where_filter(filter_input):
+        where_filters = []
+        if (
+            list(filter_input.keys())[0] == "and"
+            or list(filter_input.keys())[0] == "or"
+        ):
+            boolean_operator = list(filter_input.keys())[0]
+            conditions = list(filter_input.values())[0]
+            for condition in conditions:
+                # Could be nested AND/OR
+                filter_data = SearchAPIQueryFilterFactory.get_condition_values(
+                    condition,
+                )
+                for condition in filter_data:
+                    where_filters.append(
+                        SearchAPIWhereFilter(
+                            field=condition[0],
+                            value=condition[1],
+                            operation=condition[2],
+                            boolean_operator=boolean_operator,
+                        ),
+                    )
+        else:
+            filter_data = SearchAPIQueryFilterFactory.get_condition_values(
+                filter_input,
+            )
+            for condition in filter_data:
+                where_filters.append(
+                    SearchAPIWhereFilter(
+                        field=condition[0], value=condition[1], operation=condition[2],
+                    ),
+                )
+        return where_filters
+
+    @staticmethod
+    def get_include_filter(filter_input):
+        query_filters = []
+        for related_model in filter_input:
+            included_entity = related_model["relation"]
+            query_filters.append(SearchAPIIncludeFilter(included_entity))
+
+            if "scope" in related_model:
+                try:
+                    where_key = list(related_model["scope"]["where"].keys(),)[0]
+
+                    if where_key == "text":
+                        # TODO - we might want to move this to the data
+                        # definitions at a later point
+                        text_operator_fields = {
+                            "datasets": ["title"],
+                            "documents": ["title", "summary"],
+                            "files": ["name"],
+                            "instrument": ["name", "facility"],
+                            "samples": ["name", "description"],
+                            "techniques": ["name"],
+                        }
```

Comment:
I agree, there will be work needed to make that dictionary work for singular and plural field names.

As for my comment regarding data definitions, my point was that the dictionary probably shouldn't be 'hardcoded' in that factory class. It should be put elsewhere - be that in the same file where the PaNOSC data model is defined (that's what I meant by "data definitions"), or in `constants.py`.

Reply:
I agree that they should not be hardcoded here. It might be best to leave it as TODO for now because there might be something nicer that we can do as part of the data models themselves which may not require singular and plural dict. I am struggling to picture how that may work at this point but it may become a bit clearer later in the project. We will just have to remember that it works for some cases as it is but not for all and revisit it at some point.